pkg/apis/kubebind/v1alpha1/permissionclaims_types.go

```diff
@@ -0,0 +1,273 @@
+	// by marking it as owned by the consumer on both sides.
+	//
+	// autoDonate is mutually exclusive with autoAdopt.
+	//
+	// +optional
+	AutoDonate bool `json:"autoDonate,omitempty"`
+
+	// onConflict determines how conflicts between objects on the consumer
+	// and provider clusters will be resolved.
+	//
+	// +optional
+	// +kubebuilder:default:={}
+	OnConflict *PermissionClaimOnConflictOptions `json:"onConflict,omitempty"`
+
+	// update lists which updates to objects on the consumer cluster are claimed.
+	// By default, the whole object is synced, but metadata is not.
+	//
+	// Note that update permissions do not imply create permissions.
+	//
+	// +optional
+	Update *PermissionClaimUpdateOptions `json:"update,omitempty"`
+}
+
+type PermissionClaimReadOptions struct {
+	// labels is a list of label key wildcard patterns that are synchronized
+	// from the consumer to the provider on consumer-owned objects.
+	//
+	// +optional
+	Labels []Matcher `json:"labels,omitempty"`
+
+	// LabelsOnProviderOwnedObjects is a list of claimed label key wildcard
+	// patterns that are synchronized from the consumer cluster to the provider
+	// on objects owned by the provider.
+	//
+	// +optional
+	LabelsOnProviderOwnedObjects []Matcher `json:"labelsOnProviderOwnedObjects,omitempty"`
+
+	// annotations is a list of claimed annotation key wildcard patterns
+	// that are synchronized from the consumer cluster to the provider on
+	// objects that are owned by the consumer.
+	//
+	// +optional
+	Annotations []Matcher `json:"annotations,omitempty"`
+
+	// overrideAnnotations is a list of claimed annotation key wildcard
+	// patterns that are synchronized from the consumer cluster to the provider
+	// on provider-owned objects.
+	//
+	// +optional
+	AnnotationsOnProviderOwnedObjects []Matcher `json:"annotationsOnProviderOwnedObjects,omitempty"`
+}
+
+type Matcher struct {
+	// pattern is a wildcard pattern that is matched against the key. This means
+	// it is either a literal string or starts or ends in `*` but is not '*'
+	// itself.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength:=1
+	Pattern string `json:"pattern,omitempty"`
+}
+
+type PermissionClaimOnConflictOptions struct {
+	// recreateWhenConsumerSideDeleted set to true (the default) means the provider will recreate the object
+	// in case the object is missing on the consumer cluster, but has been synchronized before.
+	//
+	// If set to false, deleted provider-owned objects get deleted on the provider cluster as well.
+	//
+	// +kubebuilder:default:=true
+	RecreateWhenConsumerSideDeleted bool `json:"recreateWhenConsumerSideDeleted"`
+}
+
+type PermissionClaimCreateOptions struct {
+	// replaceExisting means that an existing object owned by the consumer will
+	// be replaced by the provider object.
+	//
+	// If not true, and a conflicting consumer object exists, it is not touched.
+	//
+	// +optional
+	ReplaceExisting bool `json:"replaceExisting,omitempty"`
```

Comment:
Same as above. We could reuse these if done right.

Reply:
what is "above"?

Reply:
I mean, can we have more of these and be exclusive? Like: 
```
CreateStrategy:
    DeleteAndCreate (some object in k8s world cant be updated)
    UpdateInplace
```

Reply:
How about onConflict.userObjectExists = Replace | Keep?